daemon: add cleanup for stale local ciliumendpoints that aren't being managed.

tommyp1ckles opened the pull request:
It's possible for CiliumEndpoints to become stale where they still reference existing Pods that are no longer being managed by Cilium.
In this scenario, the operator will not GC these CEPs as they have a valid pod owner reference.
This commit adds an init cleanup which cleans up stale ceps. As well, cep/ces K8s watchers will mark such CEPs for deletion and a controller GC routine will periodically GC the old CEPs.

Fixes #17631

Background, we've seen some instances where Pods CEP have become stale & out-of-sync with the actual Pod they're meant to be managing. Particularly in the following two cases:

1. Pods somehow un-managed, while retaining their CEP. One way that this can happen (i.e. I reproduced by) if the /etc/cni... Cilium config files get removed and the Node is restarted and loses its Cilium bpf state. In this case the same Pod might get re-sandboxed with another CNI (i.e. if the containers have to be restarted). At this point you'll have a Pod with a CEP but no endpoint in the endpoint manager. In this situation the CEP IP and actual Pod IP are likely to differ since the Pod has been restarted under a different CNI. Controller will not GC as the Pod.UUID and owner reference have not changed.

2. Pod becomes un-managed due to lost state. This can happen if the bpf state gets deleted for an endpoint (such as with a temporary fs following a reboot). Once you restart the Cilium pod, upon restore the existing endpoint will not be restored. But, the Pod is still running with all the Cilium state still intact.

In both cases, the agent can determine if the CEP should be deleted by checking against it's managed endpoints. If none exist then we know that that Pod is unmanaged. Endpoints that are changed, such as in the case of a Pod container being killed will have its CiliumEndpoint eventually resynced via the k8s sync controller. 

```release-note
Added Agent init check that removes all CiliumEndpoints referencing local Node that are not managed. This fixes issues where sometimes CiliumEndpoints referencing still running Pods can become unmanaged during Cilium restart.
```

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,112 @@
+// Pods that are not being managed (i.e. do not have Endpoints).
+// This must only be run after K8s Pod and CES/CEP caches are synced and local endpoint restoration is complete.
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps endpointCache, ciliumClient ciliumv2.CiliumV2Interface, cepCRDEnabled bool) error {
+	keyFn := func(ns string, name string) string { return fmt.Sprintf("%s/%s", ns, name) }
```
```suggestion
	keyFn := func(ns string, name string) string { return ns + "/" + name }
```

No need for `fmt.Sprintf`

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,112 @@
+			for i := range ces.Endpoints {
+				cep := ces.Endpoints[i]
+				if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+					cesContainedCEPLookup[fmt.Sprintf("%s/%s", ces.Namespace, cep.Name)] = &cep
```
Why not use `keyFn` from above here? I see a few instances that could benefit from this also

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,112 @@
+		// This may occur for various reasons:
+		// * Pod was restarted while Cilium was not running (likely prior to CNI conf being installed).
+		// * Local endpoint was deleted (i.e. due to reboot + temporary filesystem) and Cilium or the Pod where restarted.
+		log.WithField("ciliumendpoint", fmt.Sprintf("%s/%s", cepNamespace, cepName)).Info("found stale ciliumendpoint for local pod that is not being managed, deleting.")
```
We usually have logfields already defined (constant strings) in `pkg/logging/logfields` so that we don't have too many varying fields in the structured logging. We can probably reuse something that denotes an endpoint here.

Worth noting that we usually do this for INFO>=, aka we don't really worry about it for DEBUG fwiw.

tommyp1ckles replied:
Changed to use seperate fields with logfields.CEPName and logfields.K8sNamespace as keys.

christarazi commented on daemon/cmd/ciliumendpoints_test.go:
```diff
@@ -0,0 +1,208 @@
+		arr = append(arr, obj)
+	}
+	return arr
+}
```
Very nit, optional: I think it's easier on the eye to define these helpers / mocks down below at the end of the file rather than at the beginning so that the more important stuff like the testing code is easier to navigate to.

tommyp1ckles replied:
that makes sense

christarazi commented on pkg/k8s/factory_functions.go:
```diff
@@ -877,6 +877,19 @@ func ObjToCiliumNode(obj interface{}) *cilium_v2.CiliumNode {
 	return nil
 }
 
+// @tom: TODO: See if theres already a function for this.
```
TODO

christarazi commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -235,3 +253,59 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+// performing periodic cleanup of local stale CEP resources.
+type ciliumEndpointManager struct {
+	*lock.RWMutex
+	ciliumEndpointsCleanupEnabled bool
```
I don't actually see a use for this variable. It's being set but it's not checked anywhere. It seems like we don't need it because `enableCiliumEndpointCleanup()` will trigger the controller anyway so no need for this var.

tommyp1ckles replied:
I have this for two reasons:

Wanted to ensure that cleanStaleCEPs was done before the GC could run on this (i.e. someone could create a race if they set the gc interval to be very small)

As well, In the case that gc-interval is set to 0, and the ongoing gc is disabled, I wanted to avoid the k8s watcher potentially indefinitely appending to the marked list that never gets sweeped.

christarazi commented on pkg/option/config.go:
```diff
@@ -1083,6 +1083,9 @@ const (
 	// of new and removed datapath devices during the agent runtime.
 	EnableRuntimeDeviceDetection = "enable-runtime-device-detection"
+
+	// LocalCiliumEndpointGCInterval
```
Could you fill out the rest of this godoc?

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,112 @@
+		// This may occur for various reasons:
+		// * Pod was restarted while Cilium was not running (likely prior to CNI conf being installed).
+		// * Local endpoint was deleted (i.e. due to reboot + temporary filesystem) and Cilium or the Pod where restarted.
+		log.WithField("ciliumendpoint", fmt.Sprintf("%s/%s", cepNamespace, cepName)).Info("found stale ciliumendpoint for local pod that is not being managed, deleting.")
```
Nit: casing here, just to be more consistent with most of the log lines in the code base. FWIW like the above comment, I'd say we tend to care more about the INFO>= levels for these things.

```
❯ rg 'log.+\.(Info|Error|Warn.+|Fatal|Debug)f?\("[a-z]' --glob '!vendor' --no-filename --count | awk '{total = total + $1}END{print total}'
183

~/code/cilium/cilium master* ⇡
❮ rg 'log.+\.(Info|Error|Warn.+|Fatal|Debug)f?\("[A-Z]' --glob '!vendor' --no-filename --count | awk '{total = total + $1}END{print total}'
828
```

tommyp1ckles replied:
ah nice one liner, yea makes sense. Also applies to fmt.Errorf calls.

christarazi reviewed (commented):
Nice work! A few comments below. Overall the approach is sound.

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/sirupsen/logrus"
```
```suggestion
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/tools/cache"
	"github.com/sirupsen/logrus"
```

christarazi reviewed (approved):
Nit on import grouping / ordering, a few other instances

jrajahalme commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,116 @@
+			cepName = cep.Name
+			cepNamespace = pod.Namespace
+		} else {
+			cepObj, exists, err := d.getStore("ciliumendpoint").GetByKey(keyFn(pod.Namespace, pod.Name))
```
could this read `cepObj, exists, err := store.GetByKey(keyFn(pod.Namespace, pod.Name))` instead?

jrajahalme replied:
could also set up the key before the if blocks as it is the same for both cases?

jrajahalme commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -235,3 +252,59 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+	defer cpm.Unlock()
+	if cpm.markedForGC == nil {
+		return nil
+	}
```
This `nil` check seems superfluous, as the loop would not be entered when the slice is `nil` (`len` of a `nil` slice is 0). This code actually depends on it already as reslicing the slice does not make it `nil` when empty?

tommyp1ckles replied:
yup, you're right 👍

jrajahalme reviewed (approved):
Minor nits only :-)

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+package cmd
```
needs copyright header

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+
+	cesContainedCEPLookup := map[string]*cilium_v2a1.CoreCiliumEndpoint{}
+	if disableCEPCRD {
+		cesObjs := d.getStore("ciliumendpointslice").List()
```
This will return a copy of all objects from the store which will cause memory spikes. We should index the endpoints per node IP instead and get all endpoints from that indexer instead. [1]

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	// GetCachedPods will only return local pods in cases where ciliumendpoint CRD is disabled.
+	pods, err := d.k8sWatcher.GetCachedPods()
```
`pods` are never written in this function so `GetCachedPods()` does not need to perform a deepcopy of the object.

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+		}
+
+		var cepName, cepNamespace string
+		store := d.getStore("ciliumendpoint")
```
we could get the store before the for loop

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+		if err := ciliumClient.CiliumEndpoints(cepNamespace).Delete(ctx, cepName, metav1.DeleteOptions{}); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
+				Error("Could not clean stale CEP")
+			return fmt.Errorf("could not clean stale CEP: %w", err)
```
The delete should be best effort so we shouldn't perform an early return in case one of the CEPs failed to be deleted.

aanm replied:
Not addressed

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,117 @@
+
+	cesContainedCEPLookup := map[string]*cilium_v2a1.CoreCiliumEndpoint{}
+	if disableCEPCRD {
+		cesObjs := d.getStore("ciliumendpointslice").List()
```
I'm a little bit confused with this `ciliumendpointslice`. The endpoint slices is a feature that is not enabled by default and is behind the `option.Config.EnableCiliumEndpointSlice` option.

tommyp1ckles replied:
I'm a bit confused as well, let me dig into how these flags work again.

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1793,6 +1793,20 @@ func runDaemon() {
+				// Note: Synchronization of endpoints to their CEPs may not be complete at this point, but we only have to
+				// know what endpoints exist post-restoration in our endpointManager cache to perform cleanup.
+				if err := d.cleanStaleCEPs(ctx, d.endpointManager, k8s.CiliumClient().CiliumV2(), option.Config.DisableCiliumEndpointCRD); err != nil {
+					log.WithError(err).Fatal("Failed to clean up stale CEPs")
```
I don't know if we should fatal here, an error yes, fatal not sure...

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -880,3 +880,24 @@ func (k *K8sWatcher) GetCachedPod(namespace, name string) (*slim_corev1.Pod, err
+		if !ok {
+			return nil, fmt.Errorf("got unexpected object from pod store")
+		}
+		pods = append(pods, pod.DeepCopy())
```
This one doesn't need to be deepcopied as long the caller of this function doesn't modify the pod.

aanm commented on pkg/k8s/watchers/endpointsynchronizer.go:
```diff
@@ -77,7 +76,7 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 	e.UpdateController(controllerName,
 		controller.ControllerParams{
-			RunInterval: 10 * time.Second,
+			RunInterval: option.Config.LocalCiliumEndpointGCInterval,
```
This change could have be part of a separate commit.

aanm commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -235,3 +252,56 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+	defer cpm.Unlock()
+	for i := len(cpm.markedForGC) - 1; i >= 0; i-- {
+		cep := cpm.markedForGC[i]
+		if err := cpm.client.CiliumEndpoints(cep.Namespace).Delete(ctx, cep.Name, metav1.DeleteOptions{}); err != nil {
```
We might be doing unnecessary API calls if the CEP is already GCed by Cilium Operator. We should check if the CEP exists in the local store and only if it doesn't then delete the object.

tommyp1ckles replied:
Yeah good point, there's still the chance that the object has already been deleted and not processed but that would catch most cases.

aanm commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -235,3 +252,56 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+	defer cpm.Unlock()
+	for i := len(cpm.markedForGC) - 1; i >= 0; i-- {
+		cep := cpm.markedForGC[i]
+		if err := cpm.client.CiliumEndpoints(cep.Namespace).Delete(ctx, cep.Name, metav1.DeleteOptions{}); err != nil {
```
Also, there's a concurrency issue. How do we guarantee that a pod with the same name didn't start in between the time the CEP was marked for deletion and the time the API call is executed.

tommyp1ckles committed:
daemon: add indexer to CES watcher which indexes local CES.

Added support for indexing informer in k8s/watchers, as well
as custom indexer func which allows maintaining index on CES's
containing local endpoints by their underlying endpoint names.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
daemon: add cleanup for stale local ciliumendpoints.

It's possible for CiliumEndpoints to become stale where they still reference
existing Pods that are no longer being managed by Cilium.
In this scenario, the operator will not GC these CEPs as they have a valid pod owner reference.
This commit adds an init cleanup which cleans up stale ceps. As well, cep/ces
K8s watchers will mark such CEPs for deletion and a controller GC routine will
periodically GC the old CEPs.

Fixes #17631

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

aanm commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -928,7 +934,35 @@ func (k *K8sWatcher) K8sEventReceived(apiResourceName, scope, action string, val
+// SetStore lets you set a named cache store, only used for testing.
+func (k *K8sWatcher) SetIndexer(name string, indexer cache.Indexer) {
+	switch name {
+	case "ciliumendpointslice":
```
I would suggest to `panic` in the default case since that is a developer's mistake.

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		var cepName, cepNamespace string
+		store := d.k8sWatcher.GetStore("ciliumendpoint")
```
This store could be retrieved outside the for-loop, there's no need to do this on every iteration.

tommyp1ckles replied:
The diff is a bit confusing, I believe it no longer is in the loop here.

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,116 @@
+		var cepName, cepNamespace string
+		store := d.k8sWatcher.GetStore("ciliumendpoint")
+		key := keyFn(pod.Namespace, pod.Name)
+		if store == nil {
```
Can you add a comment here explaining why the store would be nil?

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		// See if local endpoint exists for this Pod.
+		podName := keyFn(cepNamespace, cepName)
```
UUIC this `podName` is definitely the same as `key` defined on line 65. So we could simplify the previous code and not use `cepName` nor `cepNamespace`.

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1133,6 +1133,10 @@ func initializeFlags() {
 
+	flags.Duration(option.LocalCiliumEndpointGCInterval, defaults.LocalCiliumEndpointGCInterval, "Kubernetes watcher will mark CiliumEndpoints running locally that do not have endpoints, this controls how often GC is run on stale CEPs (0 will disable CEP GC)")
+	flags.MarkHidden(option.LocalCiliumEndpointGCInterval)
+	Vp.BindEnv(option.LocalCiliumEndpointGCInterval)
```
I think this is meant to be `regOpts`, not `Vp`

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,73 @@
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps localEndpointCache, ciliumClient ciliumv2.CiliumV2Interface, enableCiliumEndpointSlice bool) error {
+	if enableCiliumEndpointSlice {
+		indexer := d.k8sWatcher.GetIndexer("ciliumendpointslice")
+		for _, cesObj := range indexer.List() {
```
I don't think this is what we want to do here. The `List()` method is part of the `Store` interface, not the `Indexer`. I guess what we want is `indexer.ByIndex("localNode", node.GetCiliumEndpointNodeIP())`

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,73 @@
+			}
+		}
+	} else {
+		store := d.k8sWatcher.GetStore("ciliumendpoint")
```
Why don't we have an indexer for "ciliumendpoint"?

tommyp1ckles replied:
This is a mistake, prior to refactoring I was listing Pods instead of CEPs directly which only watch for local Pods.

I'll add a similar indexer to the CES one.

aanm commented on daemon/cmd/ciliumendpoints_test.go:
```diff
@@ -0,0 +1,167 @@
+			managedEndpoints:   map[string]*endpoint.Endpoint{"y/foo": {}},
+			expectedDeletedSet: []string{"x/foo"},
+		},
+		"CEPs with local pods with endpoints should be GCd": {
```
`should NOT be GCd`?

aanm reviewed (changes requested):
Assuming that we have node A that gets restarted and gets a different `NodeIP`. What happens to stale `CiliumEndpoints` that were running on node A? Should we perhaps check for the node name instead? The name is static, the IPs might be ephemeral.

I do realize that we don't store the node name in the CiliumEndpoint. Perhaps we should. What do you think @joestringer? It would probably be a solution to this.

aanm commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -166,6 +168,9 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 	}
 
+	// If endpoint is local, check if it is being managed. Otherwise mark for deletion.
+	k.CheckCEP(endpoint)
```
[0]

aanm commented on pkg/k8s/watchers/cilium_endpoint_cleanup.go:
```diff
@@ -0,0 +1,127 @@
+			_, exists, err := k.ciliumEndpointStore.GetByKey(key)
+			k.ciliumEndpointStoreMU.RUnlock()
+			if err != nil {
+				return fmt.Errorf("could not get local ciliumendpoint by key from indexer: %w", err)
```
this is not an indexer, it's a store.

aanm commented on pkg/k8s/watchers/cilium_endpoint_cleanup.go:
```diff
@@ -0,0 +1,127 @@
+			log.WithField("ciliumEndpoint", key).Debug("deleting stale CiliumEndpoint")
+			k.ciliumEndpointCleanupMU.Lock()
+			defer k.ciliumEndpointCleanupMU.Unlock()
+			if err := k.client.CiliumEndpoints(cep.Namespace).Delete(ctx, cep.Name, meta_v1.DeleteOptions{}); err != nil {
```
This is a little dangerous. In statefulsets, pods will have the same name even if they are deleted and recreated. Thus, there could a race condition where:
1. The pod is deleted,
2. The CEP is already deleted but the store is not up to date yet,
3. The local endpoint was also gone
4. Now we are about to delete the CiliumEndpoint with `cep.Namespace/cep.Name`.
5. However, between step 3 and step 4, the pod was re-created as well as the CEP. Now, we will be accidentally deleting a valid CEP here.

I don't think we need the logic in this file because [Cilium Operator is GC CEPs](https://github.com/cilium/cilium/blob/70a531fe2597cf97a61148c72b463a93c871da84/operator/cmd/k8s_cep_gc.go#L25-L35) and [Cilium deletes the CiliumEndpoint when the local endpoint is gone](https://github.com/cilium/cilium/blob/bd3ad0f1e91b617e8b31ca704fc716e661739809/pkg/k8s/watchers/endpointsynchronizer.go#L382-L390). We only need to check and cleanup stale CEPs when the agent starts.

tommyp1ckles replied:
> This is a little dangerous. In statefulsets, pods will have the same name even if they are deleted and recreated. Thus, there could a race condition where:

Good point, I think this could be fixed by just adding a unique id. 

> I don't think we need the logic in this file because [Cilium Operator is GC CEPs](https://github.com/cilium/cilium/blob/70a531fe2597cf97a61148c72b463a93c871da84/operator/cmd/k8s_cep_gc.go#L25-L35) and [Cilium deletes the CiliumEndpoint when the local endpoint is gone](https://github.com/cilium/cilium/blob/bd3ad0f1e91b617e8b31ca704fc716e661739809/pkg/k8s/watchers/endpointsynchronizer.go#L382-L390). We only need to check and cleanup stale CEPs when the agent starts.

This covers any cases I can think of. However, for the endpoint sync delete to happen the Agent would've had to have been managing the endpoint at some point. If a CEP update/create causes it to get into that state Cilium won't know to clean it up. However, right now the only way for that to happen is with a manual update/create which may not be something we need to correct.

Removed the periodic gc cleanup in latest commit, will squash down later.

aanm commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -928,7 +945,35 @@ func (k *K8sWatcher) K8sEventReceived(apiResourceName, scope, action string, val
+	}
+}
+
+// SetStore lets you set a named cache store, only used for testing.
```
s/SetStore/SetIndexer

aanm commented on pkg/k8s/watchers/cilium_endpoint_slice.go:
```diff
@@ -58,7 +81,13 @@ func (k *K8sWatcher) ciliumEndpointSliceInit(client *k8s.K8sCiliumClient, asyncC
+			},
 		)
+		k.ciliumEndpointSliceStoreMU.Lock()
+		k.ciliumEndpointSliceStore = cesIndexer
```
Might worth renaming this `ciliumEndpointSliceStore` to `ciliumEndpointSliceIndexer`?

tommyp1ckles committed:
pkg/option: add flag for toggling stale CEP cleanup.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

christarazi commented on pkg/k8s/watchers/cilium_endpoint_slice.go:
```diff
@@ -22,6 +25,34 @@ var (
+				indices = append(indices, ep.Networking.NodeIP)
+			} else {
+				if ep.Networking.NodeIP == nodeIP {
+					// If we're only indexing a particular nodeIP, then as soon as we
```
Nit: 

```
else if ... {
```

rather than another indented block.

tommyp1ckles replied:
not sure what commit this was in reference to

aanm replied:
@tommyp1ckles doesn't seem resolved. This refers to the second commit.

tommyp1ckles replied:
@aanm right thanks, I need to squash down some changes I think, fixing this just gets clobbered by a conflict down the line.

christarazi commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -948,7 +954,35 @@ func (k *K8sWatcher) K8sEventReceived(apiResourceName, scope, action string, val
+		k.ciliumEndpointSliceStoreMU.Lock()
+		defer k.ciliumEndpointSliceStoreMU.Unlock()
+		return k.ciliumEndpointSliceStore
+	default:
+		panic("no such indexer: " + name)
```
Shouldn't the other similar functions also panic for the same condition now?

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,81 @@
+
+			if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+				d.checkCiliumEndpoint(ctx, cep.Namespace, cep.Name, ciliumClient, eps)
+			}
```
This code is pretty much the same except for the core `for` loop. Could we do some DRY?

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,81 @@
+
+func (d *Daemon) checkCiliumEndpoint(ctx context.Context, cepNamespace, cepName string, ciliumClient ciliumv2.CiliumV2Interface, eps localEndpointCache) {
+	keyFn := func(ns string, name string) string { return ns + "/" + name }
+	if eps.LookupPodName(keyFn(cepNamespace, cepName)) == nil {
```
Is it necessary to define `keyFn` is it's just used once in the code? `cepNamespace + '/' + cepName` seems sufficient to me.

christarazi commented on daemon/cmd/ciliumendpoints_test.go:
```diff
@@ -0,0 +1,167 @@
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+func TestCleanStaleCeps(t *testing.T) {
```
Nit: `TestCleanStaleCeps`. Let's try to keep references to CEP acronym consistent in function or type names according to https://github.com/golang/go/wiki/CodeReviewComments#initialisms. This makes grepping around the code base easier, searching with git history via `git log -S ...`, etc.

christarazi commented on daemon/cmd/daemon_test.go:
```diff
@@ -139,8 +139,12 @@ func (ds *DaemonSuite) SetUpTest(c *C) {
+	)
+	if err != nil {
+		panic("could not create daemon: " + err.Error())
+	}
```
Why is this change necessary?

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -1053,6 +1053,10 @@ func initializeFlags() {
+	flags.Bool(option.EnableStaleCiliumEndpointSliceCleanup, true, "Enables running cleanup init procedure of local CiliumEndpoints which are not being managed.")
+	flags.MarkHidden(option.EnableStaleCiliumEndpointSliceCleanup)
+	option.BindEnv(Vp, option.EnableStaleCiliumEndpointSliceCleanup)
+
```
Shouldn't the flag simply be called `EnableStaleCiliumEndpointCleanup`? Why the reference to "slice"? It makes it sound like it's specific to CES and not CEPs.

tommyp1ckles replied:
weird, I thought I changed that, I got myself into a weird git rebase mess last night and must have lost the change somehow. Ill fix it again.

christarazi reviewed (changes requested):
Forgot to click "request changes" when I submitted my review earlier.

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1721,7 +1721,7 @@ func runDaemon(ctx context.Context, cleaner *daemonCleanup, shutdowner hive.Shut
 			}
 
-			if k8s.IsEnabled() {
+			if k8s.IsEnabled() && option.Config.EnableStaleCiliumEndpointSliceCleanup {
```
CC @squeed Added this in light of CEP discussions.

christarazi commented on pkg/option/config.go:
```diff
@@ -1097,6 +1097,10 @@ const (
+
+	// EnableStaleCiliumEndpointCleanup sets whether Cilium should perform cleanup of
+	// stale CiliumEndpoints during init.
+	EnableStaleCiliumEndpointCleanup = "enable-stale-cilium-endpoint-slice-cleanup"
```
The name of the variable was fixed, but not the actual flag name :)

aanm commented on pkg/k8s/watchers/cilium_endpoint_slice.go:
```diff
@@ -58,7 +82,13 @@ func (k *K8sWatcher) ciliumEndpointSliceInit(client *k8s.K8sCiliumClient, asyncC
+				"localNode": CreateCiliumEndpointSliceLocalPodIndexFunc(),
+			},
 		)
+		k.ciliumEndpointSliceIndexerMU.Lock()
+		k.ciliumEndpointSliceIndexer = cesIndexer
+		k.ciliumEndpointSliceIndexerMU.Unlock()
```
Why don't we use the `k.SetIndexer` function here?

tommyp1ckles replied:
consistency, its done like that for every resource type: https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L129. SetStore had a note it was only for testing, but it didn't appear to be used at all so I removed it.

aanm commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -74,7 +75,13 @@ func (k *K8sWatcher) ciliumEndpointsInit(ciliumNPClient *k8s.K8sCiliumClient, as
+				"localNode": CreateCiliumEndpointLocalPodIndexFunc(),
+			},
 		)
+		k.ciliumEndpointIndexerMU.Lock()
+		k.ciliumEndpointIndexer = cepIndexer
+		k.ciliumEndpointIndexerMU.Unlock()
```
Why don't we use the `k.SetIndexer` here?

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,79 @@
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpointslice store: %T", cesObj)
+			}
+			for i := range ces.Endpoints {
```
We could do `for _, cep := range ces.Endpoints`)

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,79 @@
+			Info("Found stale ciliumendpoint for local pod that is not being managed, deleting.")
+		if err := ciliumClient.CiliumEndpoints(cepNamespace).Delete(ctx, cepName, metav1.DeleteOptions{}); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
+				Error("Could not clean stale CEP")
```
What if there is a CEP with the same name created on a different node?

tommyp1ckles replied:
If the CEP was created on a different Node but is referencing this node with NodeIP then its still stale and it will and should be deleted.

However, I think a better question is what if a stale CEP is deleted and recreated by a different Node between the check and delete. It's possible this could delete a valid CEP (similar to the EP sync race condition problem).

I think the solution is to add the UID to the delete precondition. That's easy enough for the CEP, but we don't store CES CEP UIDs in the CES store, which makes this a bit tricky.

tommyp1ckles replied:
@aanm  The never ending saga of CEP ownership issues, I think the last commit should adequately address this issue at the very rare cost of an extra API call when CES is enabled:

https://github.com/cilium/cilium/pull/20350/commits/621a1004914ff743d1b9f3b80e0deca3870de6fc

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,107 @@
+	return nil
+}
+
+func (d *Daemon) checkCiliumEndpoint(
```
The name of this function doesn't just do checking of the CEP, but it also might delete. We should clarify this and add a godoc as well, because it's definitely a non-trivial function.

christarazi replied:
It seems like we can move the `if eps.LookupPodName(cepNamespace+"/"+cepName) == nil {` condition to the caller and then that would allow us to call this function `deleteStaleCEP`.

tommyp1ckles replied:
good idea

tommyp1ckles replied:
refactored and added better docs

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,107 @@
+	// This may occur for various reasons:
+	// * Pod was restarted while Cilium was not running (likely prior to CNI conf being installed).
+	// * Local endpoint was deleted (i.e. due to reboot + temporary filesystem) and Cilium or the Pod where restarted.
+	log.WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
+		Info("Found stale ciliumendpoint for local pod that is not being managed, deleting.")
```
```suggestion
	log.WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
		Info("Found stale CEP for local pod that is not being managed, deleting.")
```

given that the other log msgs in the function already use "CEP".

qmonnet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1053,6 +1053,10 @@ func initializeFlags() {
 	flags.Bool(option.EnableBGPControlPlane, false, "Enable the BGP control plane.")
 	option.BindEnv(Vp, option.EnableBGPControlPlane)
 
+	flags.Bool(option.EnableStaleCiliumEndpointCleanup, true, "Enables running cleanup init procedure of local CiliumEndpoints which are not being managed.")
```
Nitpick:

```suggestion
	flags.Bool(option.EnableStaleCiliumEndpointCleanup, true, "Enable running cleanup init procedure of local CiliumEndpoints which are not being managed")
```

qmonnet commented on pkg/k8s/watchers/cilium_endpoint_slice.go:
```diff
@@ -22,6 +25,34 @@ var (
 )
 
+// CreateLocalNodeIndexFunc returns an IndexFunc that can index CiliumEndpointSlice objects
+// by the specified nodeIP (i.e. you can use this maintain an index of CES with endpoints
```
```suggestion
// by the specified nodeIP (i.e. you can use this to maintain an index of CES with endpoints
```

aanm replied:
@tommyp1ckles doesn't seem resolved.

tommyp1ckles replied:
my bad, fixing now

qmonnet commented on pkg/k8s/watchers/cilium_endpoint_slice.go:
```diff
@@ -30,7 +61,8 @@ func (k *K8sWatcher) ciliumEndpointSliceInit(client *k8s.K8sCiliumClient, asyncC
 
 	for {
-		_, cesInformer := informer.NewInformer(
+		// note: cesStore is has an index on node ips.
```
```suggestion
		// note: cesStore has an index on node IPs.
```

But is this comment up-to-date? I cannot find any mention of a `cesStore` in the code

qmonnet commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -948,7 +955,45 @@ func (k *K8sWatcher) K8sEventReceived(apiResourceName, scope, action string, val
+
 // GetStore returns the k8s cache store for the given resource name.
+// It's possible for valid resource names to return nil stores if that
+// watcher is not in user.
```
```suggestion
// watcher is not in use.
```

qmonnet commented on pkg/option/config.go:
```diff
@@ -2271,7 +2271,6 @@ var (
 		EnableIPv4:                   defaults.EnableIPv4,
 		EnableIPv6:                   defaults.EnableIPv6,
 		EnableIPv6NDP:                defaults.EnableIPv6NDP,
-		EnableSCTP:                   defaults.EnableSCTP,
```
Looks unrelated to this PR, was this committed by mistake?

tommyp1ckles replied:
🤔

tommyp1ckles replied:
strange, yes it seems so.

qmonnet replied:
The change is still present in the PR.

tommyp1ckles replied:
woops, fixed/rebased into wrong commit, trying again

tommyp1ckles replied:
ok that should do it.

qmonnet commented on daemon/cmd/ciliumendpoints_test.go:
```diff
@@ -0,0 +1,218 @@
+		// expectedDeletedSet contains CiliumEndpoints that are expected to be deleted
+		// during test, in the form '<namespace>/<cilium_endpoint>'.
+		expectedDeletedSet []string
+		// apiserverCEPs is used to mock out apiserver gets when running with CES enabled.
```
s/`mock out`/`mock what`/?

tommyp1ckles replied:
reworded a bit to make it clearer

qmonnet reviewed (changes requested):
> @qmonnet Still need reviews :pray:

Not sure why docs-structure was pulled on this PR, really :sweat_smile: But since we're at it: no documentation for the feature? I suppose it's because it's a hidden flag. What was the motivation for hiding it? (Sorry if it was discussed before, there are just too many resolved conversations to unfold).

I spotted a few typos/minor mistakes, please see inline below. Also one change that seems unrelated to the PR, please double-check.

The code itself looks all good as far as I can tell, nice work!

qmonnet reviewed (approved):
Looks all good to me know! Thanks for addressing the feedback, and for the explanations.

Note you've got conflicts following the merge of https://github.com/cilium/cilium/issues/21795, though.

squeed commented on daemon/cmd/daemon_main.go:
```diff
@@ -1059,6 +1059,10 @@ func initializeFlags() {
 	flags.Bool(option.EnablePMTUDiscovery, false, "Enable path MTU discovery to send ICMP fragmentation-needed replies to the client")
 	option.BindEnv(Vp, option.EnablePMTUDiscovery)
 
+	flags.Bool(option.EnableStaleCiliumEndpointCleanup, true, "Enable running cleanup init procedure of local CiliumEndpoints which are not being managed.")
```
You're the best, thanks :-)

aanm commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -0,0 +1,107 @@
+	// referenced CEP from apiserver to verify that it still references this node.
+	// To avoid excessive api calls, we only do this if CES is enabled and the CEP
+	// appears to be stale.
+	if cepUID == nil {
```
Can we add a "&& endpointSliceEnabled" just to make it more obvious that this code will only ran when CES is enabled? For the case where CES is disabled we have access to the CiliumEndpoint local cache so we can avoid doing API requests.

tommyp1ckles replied:
yeah makes sense, added with latest commit.

tommyp1ckles committed:
daemon/cmd: make CES cleanup behaviour explicit.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
k8s/watchers: prevent panic when cep has no network status.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
daemon/cmd: cleanup, remove superfluous sprintf.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

christarazi reviewed (changes requested):
Requesting until above comment is resolved.
